docs/spec/concepts.rst

```diff
@@ -3,97 +3,174 @@
+order. Static types can also be specified in other ways. For example,
+:ref:`Protocols` specify a static type which is the set of all objects which
+share a certain set of attributes and/or methods.
+
+The dynamic type
+~~~~~~~~~~~~~~~~
+
+The dynamic type :ref:`Any` represents an unknown static type. It denotes some
+unknown set of values.
+
+At first glance, this may appear similar to the static type ``object``, which
+represents the set of all Python objects. But it is quite different.
+
+If a value has the static type ``object``, a static type checker should ensure
+that all operations on the value are valid for all Python objects, or else emit
+a static type error. This allows very few operations! For example, if ``x`` is
+typed as ``object``, ``x.foo`` should be a static type error, because not all
+Python objects have an attribute ``foo``.
+
+A value typed as :ref:`Any`, on the other hand, should be assumed to have
+_some_ specific static type, but _which_ static type is not known. A static
+checker should not emit any errors that depend on assuming a particular static
+type; a static checker should instead assume that the runtime is responsible
+for checking the type of operations on this value, as in a dynamically-typed
+language.
+
+The subtype relation
+--------------------
+
+A static type ``B`` is a **subtype** of another static type ``A`` if (and only
+if) the set of values represented by ``B`` is a subset of the set of values
+represented by ``A``. Because the subset relation on sets is transitive and
+reflexive, the subtype relation is also transitive (if ``A`` is a subtype of
+``B`` and ``B`` is a subtype of ``C``, then ``A`` is a subtype of ``C``) and
+reflexive (``A`` is always a subtype of ``A``).
+
+The **supertype** relation is the inverse of subtype: ``A`` is a supertype of
+``B`` if (and only if) ``B`` is a subtype of ``A``; or equivalently, if (and
+only if) the set of values represented by ``A`` is a superset of the values
+represented by ``B``. The supertype relation is also transitive and reflexive.
+
+We also define an **equivalence** relation on static types: the types ``A`` and
+``B`` are equivalent (or "the same type") if (and only if) ``A`` is a subtype
+of ``B`` and ``B`` is a subtype of ``A``. This means that the set of values
+represented by ``A`` is both a superset and a subset of the values represented
+by ``B``, meaning ``A`` and ``B`` must represent the same set of values.
+
+We may describe a type ``B`` as "narrower" than a type ``A`` if ``B`` is a
+subtype of ``A`` and ``B`` is not equivalent to ``A``.
+
+The consistency relation
+------------------------
+
+Since :ref:`Any` (the dynamic type) represents an unknown static type, it does
+not represent any known single set of values, and thus it is not in the domain
+of the subtype, supertype, or equivalence relations on static types described
+above.
+
+We define a **materialization** relation from gradual types to gradual and
```

Comment:
Let's give an intuitive explanation of what materialization is: if B is a materialization of A then B is a "more static" version of A, A is a "more dynamic" version of B.

Also, let's use a convention that letters closer to the beginning of the alphabet are either generally more static or generally more dynamic (it helps the reader to keep the directions of things straight, especially with things like contravariance).

I think it's important to highlight that different occurrences of Any can be replaced with different types.  We could do that by adding "... replacing ... with some gradual or static type, possibly different for each occurrence, results in...".

Another possibility is to define a single step of replacement and make materialization the transitive closure of that relation.  Then you can get from `tuple[Any, Any]` to `tuple[int, str]` in two steps of replacement.  I'm not sure if that's better.

Also, this relation is almost reflexive (and maybe we want it to be actually reflexive).  The definition written here allows replacing one occurrence of `Any` with `Any` so gradual types are materializations of themselves.  But this definition doesn't make static types materializations of themselves.  That could be repaired by saying "zero or more" and making this a relation on all types, or with the single step of replacement and saying materialization is the reflexive transitive closure.

Reply:
I think I'm happy enough with fixing this with your "zero or more" suggestion, but open to explicitly adding the single-step relation if it looks preferable to you after seeing my update.